### DISCUSSION TITLE
Is there a way to run next.js without a server when not doing SSR at all?

### DISCUSSION THREAD

**Discussion by Albert-Gao** [vercel/next.js#41093] [Help]
- I want to build a SPA for a SaaS
- doing SSG for the non-auth pages
- for the 99% part which is auth-protected pages,  I want to build as a SPA, and not run on a server
- which means, the whole site is generated like an app created by CRA, ONLY containing static assets

Is it possible for next.js? might know the answer is no, but still trying my luck here.

thanks :)

**Comment by gaston-flores** [comment]
If you don't need a server at all, you can use the [static export feature](https://nextjs.org/docs/advanced-features/static-html-export) to generate static HTML files. Keep in mind, [many features in next require a server to work.](https://nextjs.org/docs/advanced-features/static-html-export#unsupported-features)

**Reply by Albert-Gao** [reply]
Thank you so much, I do not need any of those features, maybe that Image optimization, but that's fine.

this is simply a god send, and I just noticed that if I static export the whole site, then getInitialProps becomes a route level prefetching!!! Which effectively enabled concurrent React! Jesus!! Did not realize this, thought next.js is only capable for doing SSR

Thanks a lot!

Just 1 quick question, for the exported version, where should I redirect the user to for any non-index page call? 
for example, in a SPA, for a page call like /user/123, if the user refreshes the browser, I would redirect the request to '/', since then the client side routing would kick in.

I did the same to the static export version, but it seems it causes some redirect loop on my server, have not found anything about it in the doc, or, i would say, how to redirect the dynamic route in a static exported version?

I found I might need to configure https://nextjs.org/docs/api-reference/next.config.js/exportPathMap, but not sure how to configure a dynamic version like "/user/:userId", any ideas? thanks

what I am missing here? thanks :)

**Reply by gaston-flores** [reply]
> Just 1 quick question, for the exported version, where should I redirect the user to for any non-index page call?
for example, in a SPA, for a page call like /user/123, if the user refreshes the browser, I would redirect the request to '/', since then the client side routing would kick in.

I think you could make use of `getStaticPaths` for this. If your `pages` directory looks like this

```
.
│   about.jsx
│   index.jsx
│   _app.jsx
│   _document.jsx
│   
└───users
        index.jsx
        [id].jsx
```

And your getStaticPaths in `[id].jsx` is

```jsx
export async function getStaticPaths() {
  return {
    paths: [{ params: { id: "1" } }, { params: { id: "2" } }],
    fallback: false,
  };
}
```

The result of running `next build` + `next export` will be

```
.
│   404.html
│   about.html
│   index.html
│   posts.html
│   
└───users
       1.html
       2.html
```

You should try to serve the file at that location first, and only fall back to the 404 or index page if there is no pre-rendered html for the uri the user requested
